### PR TITLE
Fix NPE thrown by prefix and regex query in strange scenarios

### LINKED ISSUE #94364: Error when doing a prefix search when the prefix is min_chars -1
### Elasticsearch Version

8.6.2

### Installed Plugins

elastiknn-8.6.2.1.zip

### Java Version

openjdk 18.0.2.1 2022-08-18

### OS Version

docker: Linux c30f7a57fb60 5.4.0-139-generic #156-Ubuntu SMP Fri Jan 20 17:27:18 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Problem Description

When doing a prefix search over a text field with index_prefixes and min_chars = X  , if the length of the prefix is X-1, the search fails with a lucene null pointer error:

this error has been reported (and solved)  for  openserarch, here https://github.com/opensearch-project/OpenSearch/issues/2826

### Steps to Reproduce

1 crate an index with index prefix and index some docs:
```  
PUT test_prefix
{
  "mappings": {
    "properties": {
      "t": {
        "type": "text",
        "index_prefixes": {"min_chars":3}
      }
    }
  }
}

POST test_prefix/_doc
{
  "t":"This is the test to these kind of errors"
}
```
Do the search , the first and second work correctly while the last one, with a prefix of min_chars-1 fails
```
GET test_prefix/_search
{"query":{"prefix":{"t": "t"}}}

GET test_prefix/_search
{"query":{"prefix":{"t": "the"}}}

GET test_prefix/_search
{"query":{"prefix":{"t": "th"}}}

```
simple_query_string also fails

```
GET test_prefix/_search
{"query":{"simple_query_string":{"fields":["t"], "query":"th*"}}}
```

The regexp is the only option to do the search, but i could not find information on the documentation if  regexp queries take advantage of the index_prefix  
```
GET test_prefix/_search
{"query":{"regexp": {
      "t": {
        "value": "th.*",
        "flags": "ALL",
        "case_insensitive": true,
        "max_determinized_states": 10000,
        "rewrite": "constant_score"
      }
    }}}
```

### Output

the error produced is the next
```json
{
  "error" : {
    "root_cause" : [
      {
        "type" : "null_pointer_exception",
        "reason" : "Cannot invoke \"org.apache.lucene.search.MultiTermQuery$RewriteMethod.rewrite(org.apache.lucene.index.IndexReader, org.apache.lucene.search.MultiTermQuery)\" because \"this.rewriteMethod\" is null"
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "test_prefix",
        "node" : "KPamBItNTOi2hLRNTFcTMw",
        "reason" : {
          "type" : "null_pointer_exception",
          "reason" : "Cannot invoke \"org.apache.lucene.search.MultiTermQuery$RewriteMethod.rewrite(org.apache.lucene.index.IndexReader, org.apache.lucene.search.MultiTermQuery)\" because \"this.rewriteMethod\" is null"
        }
      }
    ],
    "caused_by" : {
      "type" : "null_pointer_exception",
      "reason" : "Cannot invoke \"org.apache.lucene.search.MultiTermQuery$RewriteMethod.rewrite(org.apache.lucene.index.IndexReader, org.apache.lucene.search.MultiTermQuery)\" because \"this.rewriteMethod\" is null",
      "caused_by" : {
        "type" : "null_pointer_exception",
        "reason" : "Cannot invoke \"org.apache.lucene.search.MultiTermQuery$RewriteMethod.rewrite(org.apache.lucene.index.IndexReader, org.apache.lucene.search.MultiTermQuery)\" because \"this.rewriteMethod\" is null"
      }
    }
  },
  "status" : 500
}

```

### PR DESCRIPTION
**Pull request by benwtrent**
In certain scenarios, running a MultiTerm query sets a `null` rewrite method. While `null` is usually checked, there are branches in the code where this is not adequately checked.

Additionally, `MultiTermQuery#setRewriteMethod` has been deprecated for a while. So, to correct this bug, 

 - Remove calls to `MultiTermQuery#setRewriteMethod` where possible
 - Always check for `null` rewrite method

closes: https://github.com/elastic/elasticsearch/issues/94364

### CODE REVIEW

**Review by romseygeek — CHANGES_REQUESTED** [review]
Can you add a test for the specific scenario that's failing in that test report?  It's a bit alarming that we don't already cover it.

Are there still some places left that call `setRewriteMethod()` after this change?

**Comment by benwtrent** [comment]
> Are there still some places left that call `setRewriteMethod()` after this change?

Two places: 

https://github.com/elastic/elasticsearch/blob/c513b2bcc68bf730aef5a13cf36737c4b38d713f/server/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java#L337-L338

https://github.com/elastic/elasticsearch/blob/edd7749164aa6873a0eb7565abcc612d29c601f0/server/src/main/java/org/elasticsearch/index/search/MatchQueryParser.java#L477-L480

Those two places seemed really weird to me. 

I think we could get around them by adding `RewriteMethod` as a parameter to `FieldType#fuzzyQuery` but that was even a bigger change. If you think that is good, I will happily do it.

> Can you add a test for the specific scenario that's failing in that test report? It's a bit alarming that we don't already cover it.

Oh yeah, I am working on a test now. Wanted to get your opinion on the current fix.

**Comment by romseygeek** [comment]
> I think we could get around them by adding RewriteMethod as a parameter to FieldType#fuzzyQuery but that was even a bigger change. If you think that is good, I will happily do it.

Let's do it.  And +1 to the direction of the PR in general.

File: modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/SearchAsYouTypeFieldMapper.java [hunk 1]
```diff
@@ -413,7 +413,9 @@ public SpanQuery spanPrefixQuery(String value, SpanMultiTermQueryWrapper.SpanRew
                 SpanMultiTermQueryWrapper<?> spanMulti = new SpanMultiTermQueryWrapper<>(
                     new PrefixQuery(new Term(name(), indexedValueForSearch(value)))
                 );
-                spanMulti.setRewriteMethod(method);
+                if (method != null) {
+                    spanMulti.setRewriteMethod(method);
```
Comment (romseygeek) [inline modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/SearchAsYouTypeFieldMapper.java:417]: I should probably go deprecate this in lucene as well, just for consistency.  For a follow-up.

File: server/src/main/java/org/elasticsearch/index/search/QueryStringQueryParser.java [hunk 2]
```diff
@@ -490,7 +489,8 @@ private Query getFuzzyQuerySingle(String field, String termStr, int minSimilarit
                 getFuzzyPrefixLength(),
                 fuzzyMaxExpansions,
                 fuzzyTranspositions,
-                context
+                context,
+                null
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/search/QueryStringQueryParser.java:493]: This should probably pass `this.fuzzyRewriteMethod` shouldn't it? But I guess that's a behaviour change - maybe add a TODO and we can call it out separately as a bugfix?

**Review by romseygeek — APPROVED** [review]
LGTM, thanks for taking care of this!

File: server/src/main/java/org/elasticsearch/index/search/QueryStringQueryParser.java (hunk 2 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/search/QueryStringQueryParser.java:493]: OK, looking at `QueryParserBase#getFuzzyQuery` (the thing this class overrides), Lucene doesn't set the `fuzzyRewriteMethod` there either. 

Do we want to do that here? I am honestly not sure.

**Pull request merged**